### ISSUE TITLE
Hex encoding not everywhere

### ISSUE THREAD

**Issue by markov2** [sysread/URI-Fast#29]
The URI RFC3968 describes how hex encoding takes place, but that does not mean that all components can have hex encoding: simply because their character-set is restricted.

During normalization, we do not need to uc_hex() the scheme, host or port.

**Comment by sysread** [comment]
I'm not entirely certain that is true for host. From RFC3986, section 3.2.2:

`Although host is case-insensitive, producers and normalizers should use lowercase for registered names and hexadecimal addresses for the sake of uniformity, while only using uppercase letters for percent-encodings.`

**Comment by sysread** [comment]
Yeah, I am pretty sure we do need it for the host. From 3.2.2:

```
   A host identified by a registered name is a sequence of characters
   usually intended for lookup within a locally defined host or service
   name registry, though the URI's scheme-specific semantics may require
   that a specific registry (or fixed name table) be used instead.  The
   most common name registry mechanism is the Domain Name System (DNS).
   A registered name intended for lookup in the DNS uses the syntax
   defined in Section 3.5 of [RFC1034] and Section 2.1 of [RFC1123].
   Such a name consists of a sequence of domain labels separated by ".",
   each domain label starting and ending with an alphanumeric character
   and possibly also containing "-" characters.  The rightmost domain
   label of a fully qualified domain name in DNS may be followed by a
   single "." and should be if it is necessary to distinguish between
   the complete domain name and some local domain.

      reg-name    = *( unreserved / pct-encoded / sub-delims )
```

**Comment by sysread** [comment]
Addressed `scheme` and `port` in 4fc4b46

**Comment by markov2** [comment]
Although the generic URI spec permits %-encoding, there are protocol specific restrictions on http[s]
Question is whether normalization require to fix all non-conformities?  As said with
```
   The reg-name syntax allows percent-encoded octets in order to
   represent non-ASCII registered names in a uniform way that is
   independent of the underlying name resolution technology.  Non-ASCII
   characters must first be encoded according to UTF-8 [STD63], and then
   each octet of the corresponding UTF-8 sequence must be percent-
   encoded to be represented as URI characters.  URI producing
   applications must not use percent-encoding in host unless it is used
   to represent a UTF-8 character sequence.  When a non-ASCII registered
   name represents an internationalized domain name intended for
   resolution via the DNS, the name must be transformed to the IDNA
   encoding [RFC3490] prior to name lookup.  URI producers should
   provide these registered names in the IDNA encoding, rather than a
   percent-encoding, if they wish to maximize interoperability with
   legacy URI resolvers.
```
So: should you fix the lack of IDNA encoding on those schemes where "host" is a DNS name?  Or do you solve that in the host() setter?  Or is URI::Fast not implementing protocol differences.  Choices.

Yes, I know... it is a lot more work to implement protocol specific normalization.  My billions of URIs do probably contain a lot of links which are bad on purpose: to get additional counts at Google to up ranking.  "URL producing applications" are not always following requirements.
